pkg/hubble/metrics/dynamic_flow_processor.go

```diff
@@ -0,0 +1,137 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package metrics
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+
+	flowpb "github.com/cilium/cilium/api/v1/flow"
+
+	"github.com/cilium/cilium/pkg/hubble/metrics/api"
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+// DynamicFlowProcessor represents instance of hubble exporter with dynamic
+// configuration reload.
+type DynamicFlowProcessor struct {
+	logger  logrus.FieldLogger
+	watcher *metricConfigWatcher
+	// Protects against deregistering metric handlers while ProcessFlow is executing, or concurrent config reloads.
+	mutex    lock.RWMutex
+	Metrics  []api.NamedHandler
+	registry *prometheus.Registry
+}
+
+// OnDecodedEvent distributes events across all managed exporters.
+func (d *DynamicFlowProcessor) OnDecodedFlow(ctx context.Context, flow *flowpb.Flow) (bool, error) {
+	select {
+	case <-ctx.Done():
+		return false, nil
+	default:
+	}
+
+	d.mutex.RLock()
+	defer d.mutex.RUnlock()
+
+	var errs error
+	if d.Metrics != nil {
+		errs = api.ExecuteAllProcessFlow(ctx, flow, &d.Metrics)
+	}
+
+	if errs != nil {
+		d.logger.WithError(errs).Error("Failed to ProcessFlow in metrics handler")
+	}
+	return false, errs
+}
+
+// NewDynamicFlowProcessor creates instance of dynamic hubble flow exporter.
+func NewDynamicFlowProcessor(reg *prometheus.Registry, logger logrus.FieldLogger, configFilePath string) *DynamicFlowProcessor {
+	dynamicFlowProcessor := &DynamicFlowProcessor{
+		logger:   logger,
+		registry: reg,
+	}
+	watcher := NewMetricConfigWatcher(configFilePath, dynamicFlowProcessor.onConfigReload)
+	dynamicFlowProcessor.watcher = watcher
+	return dynamicFlowProcessor
+}
+
+func (d *DynamicFlowProcessor) onConfigReload(ctx context.Context, isSameHash bool, hash uint64, config api.Config) {
+	if isSameHash {
+		return
+	}
```

Comment:
This could probably be moved into the caller and the argument can be removed.

Reply:
done